File: Prácticas/Deepspace_P3/lib/Weapon.rb
#encoding:utf-8

require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace

# Class to represent the weapons that a space station can dispose of to
# increment its energy when shooting
#
# @author Miguel Ángel Fernández Gutiérrez
class Weapon

	# Initializers
	# ==========================================================================

	# Class initializer
	# @param _name [String] name of weapon
	# @param _type [WeaponType] type of weapon
	# @param _uses [Integer] number of uses of weapon
	def initialize(_name, _type, _uses)
		# @!attribute [String] name of weapon
		@name = _name

		# @!attribute [WeaponType] type of weapon
		@type = _type

		# @!attribute [Intenger] number of uses of weapon
		@uses = _uses
	end

	# Copy constructor
	# @param origin [Weapon] instance which is going to be copied
	# @return [Weapon] a copy of the given instance
	def self.newCopy(origin)
		return new(origin.name, origin.type, origin.uses)
	end

	# Getters
	# ==========================================================================

	attr_reader :type, :uses, :name

	# Get type's power
	# @return [Float] type's power
	def power
		return @type.power
	end

	# Setters
	# ==========================================================================
	
	# Uses the weapon
	# If it is still available (uses is greater than 0), we can use the weapon
	# Otherwise, we cannot use the weapon
	# @return [Float] boost if uses > 0; 1.0 otherwise
	def useIt
		if @uses > 0
			@uses = @uses - 1
			return power
		else
			return 1.0
		end
	end

	# String representation, UI version
	# ==========================================================================

	# String representation of the object
	# @return [String] string representation
	def to_s
		message = "[Weapon]-> Name: #{@name}, Type: #{@type}, Power: #{power}, " \
				+ "Uses: #{@uses}"
        return message
	end

	# To UI
	def getUIversion
		return WeaponToUI.new(self)
	end
end # class Weapon

end # module Deepspace

File: Prácticas/deepspace/src/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

/**
 * Class to represent the weapons that a space station can dispose of to
 * increment its energy when shooting
 * 
 * @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
 */
class Weapon implements CombatElement {
    /**
     * Name of weapon.
     */
    private String name;
    
    /**
     * Type of weapon.
     */
    private WeaponType type;
    
    /**
     * Number of uses of weapon.
     */
    private int uses;
    
    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------
    
    /**
     * Class initializer.
     * @param _name name of weapon
     * @param _type type of weapon
     * @param _uses number of uses of weapon
     */
    Weapon(String _name, WeaponType _type, int _uses) {
        name = _name;
        type = _type;
        uses = _uses;
    }
    
    /**
     * Copy constructor.
     * @param w instance which is going to be copied
     */
    Weapon(Weapon w) {
        this.name = w.name;
        this.type = w.type;
        this.uses = w.uses;
    }
    
    // -------------------------------------------------------------------------
    // Getters
    // -------------------------------------------------------------------------
    
    /**
     * Getter for type.
     * @return type
     */
    public WeaponType getType() {
        return type;
    }
    
    /**
     * Getter for uses.
     * @return uses
     */
    public int getUses() {
        return uses;
    }
    
    /**
     * Get type's power.
     * @return type's power
     */
    public float power() {
        return type.getPower();
    }
    
    // -------------------------------------------------------------------------
    // Setters
    // -------------------------------------------------------------------------
    
    /**
     * Uses the weapon.
     * If it is still available (uses is greater than 0), we can use the weapon
     * Otherwise, we cannot use the weapon
     * @return boost if uses > 0, 1.0 if uses == 0
     */
    public float useIt() {
        if ( uses > 0 ) {
            uses--;
            return power();
        } else
            return (float) 1.0;
    }
    
    // -------------------------------------------------------------------------
    // String representation, UI version
    // -------------------------------------------------------------------------
    
    /**
     * String representation of the object.
     * @return string representation
     */
    public String toString() {
        String message = "[Weapon] -> Name: " + name
                + ", Type: " + type
                + ", Power: " + power()
                + ", Uses: " + uses;
        return message;
    }
    
    /**
     * To UI.
     */
    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }
}
